Replace EVP_MAC_CTX_copy() by EVP_MAC_CTX_dup()

kroeckx opened the pull request:
- [ ] documentation is added or updated

levitte commented on crypto/evp/pkey_mac.c:
```diff
@@ -250,10 +248,12 @@ static int pkey_mac_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
-                                     (EVP_MAC_CTX *)ctx->pkey->pkey.ptr))
-                return 0;
+            if (ctx->pkey != NULL) {
+                /* Can leak memory? */
```
Yup. That's what happens with the *one* CMAC test that has Travis fail. You must free `hctx->ctx` before assigning the new duplicate.

kroeckx replied:
It's it great that can add a comment at the exact place where you're leaking. I didn't fully understand part of the code at that point in time.

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -48,18 +48,19 @@ void EVP_MAC_CTX_free(EVP_MAC_CTX *ctx)
-    macdata = dst->data;
-    *dst = *src;
-    dst->data = macdata;
+    if (src->data != NULL && src->meth->copy(dst->data, src->data) == 0) {
```
If you want to do this, don't do it half way, change the method function too.

kroeckx replied:
I'm not sure what you mean. I have an other commit that changes the copy function to a dup function, it's just not pushed yet.

levitte replied:
Oh good, that is exactly what I was asking for

kroeckx replied:
Just pushed it

levitte replied:
Uhmmmm, this is still just half way?

I do see one instance where you use the copy method function directly, will comment there.

kroeckx commented:
I guess I should already rebase now fix to fix the gmac code too.

kroeckx commented:
So I think part of the commit is wrong now, the evp_mac_st function is a copy function currently, not a dup function. Not sure what I'm going to do with it yet.

levitte commented:
It becomes evident that using dup semantics instead of copy semantics can't easily be an isolated thing. It's part of a chain of calls, and if you really want to do this, you either have to find a good spot to have copy semantics converted to dup semantics (which you currently seem to do in `pkey_mac.c`), or change copy to dup all the way. And yeah, the former is going to feel clumsy, there's no way around that.

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -48,18 +48,19 @@ void EVP_MAC_CTX_free(EVP_MAC_CTX *ctx)
-    macdata = dst->data;
-    *dst = *src;
-    dst->data = macdata;
+    if (src->data != NULL && src->meth->dup(dst->data, src->data) == 0) {
```
Because EVP_MAC is a new type, I think you can freely change the dup function to be a real dup

levitte commented on crypto/evp/pkey_mac.c:
```diff
@@ -81,7 +82,8 @@ static int pkey_mac_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
 
-    if (!EVP_MAC_CTX_copy(dctx->ctx, sctx->ctx))
+    if (sctx->ctx->data != NULL
+        && !dctx->ctx->meth->dup(dctx->ctx->data, sctx->ctx->data))
```
Shouldn't this be this?

``` C
    EVP_MAC_CTX_free(dctx->ctx);
    if ((dctx->ctx = EVP_MAC_CTX_dup(sctx->ctx)) == NULL) 
```

kroeckx replied:
That dctx->ctx->meth->dup() function is really a copy function that I renamed by mistake. But that would work too.

kroeckx replied:
The pkey_mac_init() above just allocated it, directly calling free on it doesn't make much sense.

levitte commented on crypto/include/internal/evp_int.h:
```diff
@@ -44,7 +44,7 @@ struct evp_pkey_method_st {
     int flags;
     int (*init) (EVP_PKEY_CTX *ctx);
-    int (*copy) (EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src);
+    int (*dup) (EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src);
```
I don't think this is a good idea unless you want to break the API.

kroeckx replied:
This is an internal function, I just change the name of the field in that internal structure to say what the functions actually do.

kroeckx commented:
I'm not sure what you mean. In EVP_PKEY_METHOD, it was called copy, but the copy functions allocated a new struct. In EVP_MAC you have copy function that really just overwrite things in an already allocated struct. I renamed both to dup, which clearly is wrong. Either the the name in EVP_MAC should stay copy, which is a 1 line revert, or the functions should get converted to dup functions.

There is an EVP_PKEY_CTX_dup() function. You introduced an EVP_MAC_CTX_copy() function which I've changed to EVP_MAC_CTX_dup(). I believe the _copy function is not something we want as external interface. It's not clearly defined on what it actually does, nor is easy to get this correct.

In case we only have an external dup function, how we split the allocation of the new EVP_MAC_IMPL and the copying of an EVP_MAC_IMPL from the source EVP_MAC_CTX to the new one doesn't really matter. The difference is just that the dup() function would internally call the init() function, while a copy function doesn't.

levitte commented:
I think you and I have different ideas of what a dup does. Your angle seems to be more about the inner workings (dup allocates the destination and copies the source into that) while mine is more about the interface (dup takes *only* source parameters and returns the newly allocated destination). From my point of view, something called dup that takes a destination parameter is confusing. That's the basis of some of my comments.

kroeckx commented:
The dup in EVP_PKEY_METHOD is a little strange. It takes 2 EVP_PKEY_CTX parameters, and then really does the effect of dst->data = dup(src->data), where data is the method specific data, for instance DH_PKEY_CTX.  But it also changed dst itself, for instance pkey_dh_init() will write dst->keygen_info.

levitte commented:
*All* the method copy functions are really just helpers to do whatever's needed to do a deep copy of the public API structure (i.e., in the EVP_PKEY_METHOD case, the EVP_PKEY_CTX structure).  If that means allocating new space or copying bits into already allocated space is, from public API point of view, less important, so it's perfectly fine if the copy is implemented as a dup of the algorithm private parts, the layer above doesn't really care.

I think that a detail that makes this confusing is that in all our EVP APIs, we pass down the full context to the algorithm implementations, which does give them freedom to do as they please...  but also means that the separation between the EVP layer and the algorithm implementations isn't very strict.
I deviated from that pattern in EVP_MAC, which does mean that we have to be much more strict in how things are done there, and is a reason why a copy method function *must* have a typical copy API and *must* behave in a typical copying manner (i.e. it can't be a dup), and why a dup method function *must* have a typical dup API and *must* behave in a typical duplicating manner.

That's why I'm trying to push you to make the interface correspond to commonly used copy or dup patterns, at least for EVP_MAC.  In other words:

- if you want a copy function, please use the signature `int ..._copy(TYPE *dst, TYPE *src)`
- if you want a dup function, please use the signature `TYPE *..._dup(TYPE *src)`

Also, the more I think of it, the more it makes sense to bridge between those two patterns in `pkey_mac.c`...  please don't go renaming method functions from copy to dup without changing that function's signature, that's simply asking for confusion.

kroeckx commented:
I've forced pushed a new commit

levitte commented on crypto/evp/pkey_mac.c:
```diff
@@ -81,7 +82,8 @@ static int pkey_mac_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
 
-    if (!EVP_MAC_CTX_copy(dctx->ctx, sctx->ctx))
+    if (sctx->ctx->data == NULL
+        || !dctx->ctx->meth->copy(dctx->ctx->data, sctx->ctx->data))
```
This seems to be the only spot, other than in `mac_lib.c`, where you use the copy method function directly.  Would you consider some form of this instead?

``` C
    EVP_MAC_CTX_free(dctx->ctx);
    if ((dctx->ctx = EVP_MAC_CTX_dup(sctx->ctx)) == NULL)
        goto err;
```

That would allow you to convert the EVP_MAC copy function to a dup function.

kroeckx replied:
I find the sequence a little strange in that case:
```
ctx = malloc();
free(ctx);
ctx = dup();
```

Which is why I didn't do that.

levitte replied:
Yup, I know.  That's the exact reason why I *didn't* implement dup semantics, because it caused this type of strangeness, since the rest of OpenSSL uses copy semantics.  That's the cost of bridging between the two ways.

Of course, we can keep it this way, but then I find it strange to see dup semantics at the EVP_MAC API level, but still have a copy function in the EVP_MAC method.  As a method author, I'd raise an eye-brow and wonder what the hell is going on.  That's also a price to pay, although different.

levitte replied:
So I guess that leaves you with a choice, and it would be good if it was documented somehow.

kroeckx replied:
The external API we are implementing here is *dup*, this function is called by EVP_PKEY_CTX_dup(). That is, we allocate a new one with init(), and the copy() it.

kroeckx replied:
Well, we don't allocate with init(), but we call it to initialize things. Which for a dup function doesn't actually make sense, since you want to overwrite whatever init() did.

levitte replied:
Oooohhhhh...  hadn't payed attention to `EVP_PKEY_CTX_dup()`, and this gets quite confusing since all *other* EVP context types use copy in their public API :-/

levitte commented on crypto/evp/pkey_mac.c:
```diff
@@ -75,15 +76,25 @@ static int pkey_mac_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
-    dctx = EVP_PKEY_CTX_get_data(dst);
+    dctx = OPENSSL_zalloc(sizeof(*dctx));
+    if (dctx == NULL) {
+        EVPerr(EVP_F_PKEY_MAC_INIT, ERR_R_MALLOC_FAILURE);
```
You need to change the first argument to `EVP_F_PKEY_MAC_COPY`

levitte commented on crypto/err/openssl.txt:
```diff
@@ -744,6 +744,7 @@ EVP_F_EVP_ENCRYPTUPDATE:167:EVP_EncryptUpdate
 EVP_F_EVP_MAC_CTRL:209:EVP_MAC_ctrl
 EVP_F_EVP_MAC_CTRL_STR:210:EVP_MAC_ctrl_str
 EVP_F_EVP_MAC_CTX_COPY:211:EVP_MAC_CTX_copy
+EVP_F_EVP_MAC_CTX_DUP:221:EVP_MAC_CTX_dup
```
You can remove the `EVP_F_EVP_MAC_CTX_COPY` line...  probably even reuse its number.

levitte commented on crypto/evp/evp_err.c:
```diff
@@ -59,6 +59,7 @@ static const ERR_STRING_DATA EVP_str_functs[] = {
     {ERR_PACK(ERR_LIB_EVP, EVP_F_EVP_MAC_CTRL, 0), "EVP_MAC_ctrl"},
     {ERR_PACK(ERR_LIB_EVP, EVP_F_EVP_MAC_CTRL_STR, 0), "EVP_MAC_ctrl_str"},
     {ERR_PACK(ERR_LIB_EVP, EVP_F_EVP_MAC_CTX_COPY, 0), "EVP_MAC_CTX_copy"},
+    {ERR_PACK(ERR_LIB_EVP, EVP_F_EVP_MAC_CTX_DUP, 0), "EVP_MAC_CTX_dup"},
```
You can remove the `EVP_F_EVP_MAC_CTX_COPY` line...

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -48,18 +48,28 @@ void EVP_MAC_CTX_free(EVP_MAC_CTX *ctx)
+        return NULL;
+    }
+
+    dst->meth = src->meth;
```
To make it future proof, I'd suggest doing `*dst = *src`.  It doesn't really matter that `dst->data` becomes a copy of `src->data` for a short moment.

levitte commented on crypto/evp/mac_lib.c:
```diff
@@ -48,18 +48,28 @@ void EVP_MAC_CTX_free(EVP_MAC_CTX *ctx)
+    if (src->data == NULL)
+        return NULL;
+
+    dst = OPENSSL_zalloc(sizeof(EVP_MAC_CTX));
```
Might I suggest `dst = OPENSSL_zalloc(sizeof(*dst));` ?

levitte commented on crypto/modes/siv128.c:
```diff
@@ -156,32 +164,40 @@ int CRYPTO_siv128_init(SIV128_CONTEXT *ctx, const unsigned char *key, int klen,
+        EVP_MAC_CTX_free(mac_ctx);
+        goto err;
     }
+    EVP_MAC_CTX_free(mac_ctx);
```
I dunno, I kinda liked to cascade flow that was there previously, so what about this?

``` C
    if (key == NULL || cbc == NULL || ctr == NULL
            || (ctx->cipher_ctx = EVP_CIPHER_CTX_new()) == NULL
            || (ctx->mac_ctx_init = EVP_MAC_CTX_new_id(EVP_MAC_CMAC)) == NULL
            || !EVP_MAC_ctrl(ctx->mac_ctx_init, EVP_MAC_CTRL_SET_CIPHER, cbc)
            || !EVP_MAC_ctrl(ctx->mac_ctx_init, EVP_MAC_CTRL_SET_KEY, key, klen)
            || !EVP_EncryptInit_ex(ctx->cipher_ctx, ctr, NULL, key + klen, NULL)
            || (mac_ctx = EVP_MAC_CTX_dup(ctx->mac_ctx_init)) == NULL
            || !EVP_MAC_update(mac_ctx, zero, sizeof(zero))
            || !EVP_MAC_final(mac_ctx, ctx->d.byte, &out_len)) {
        EVP_CIPHER_CTX_free(ctx->cipher_ctx);
        EVP_MAC_CTX_free(ctx->mac_ctx_init);
        EVP_MAC_CTX_free(mac_ctx);
        return 0;
    }
    EVP_MAC_CTX_free(mac_ctx);
```

levitte commented on doc/man3/EVP_MAC.pod:
```diff
@@ -344,6 +338,10 @@ look like this:
 
+=head1 HISTORY
+
+The functions are added in OpenSSL 3.0.0.
```
`These functions were added in `...

At the time of release, past tense will make sense.

levitte commented:
All in all, I like what I see at this point.

kroeckx commented:
I've pushed a new version.

levitte commented:
Travis is having some odd failure...  it's unrelated to this PR, though.

kroeckx commented:
So are you  happy with the current version?

kroeckx commented:
@levitte: Do you want to review this?

kroeckx commented:
@openssl/committers: I'm still waiting for someone to review this.

kroeckx commented:
So I've rebased it. There were conflicts in tls1_prf.c and siv128.c, and new code in sskdf.c.

Can you reconfirm?

t8m commented:
The travis failure seems relevant? EVP function code 229 collision at EVP_F_PKEY_MAC_COPY

t8m commented:
The Travis indicates there is a leak in evp_kdf_test and evp_test with KDF vectors. I do not know whether it is related but it could be.

kroeckx commented:
I've rebased it again to fix a doc-nits problem that was fixed in master.

t8m commented:
It still seems to break something - this time enable-ssl3 option.

kroeckx commented:
That doesn't look like a regression with this patch, master is broken in the same way.

t8m commented:
Ah, right, I looked at some Travis run on master that passed but that did not include extended tests.

t8m commented:
So I am approving this again.
